Docs: Filter and group by public preview updates (#122304)
grafana/grafana: docs/sources/visualizations/dashboards/build-dashboards/create-dashboard/dashboard-controls.md

# Dashboard controls

> In the **Dashboard controls** section of the sidebar, you can add variables, annotation queries, dashboard links, and controls to filter and group data, without leaving the dashboard.

Reviewer: I think I understand why you moved the filter to the end but if we wanted to keep it in order it apperas in the UI, filter and group by should go first

also, should this be `add`?
`add controls to filter and group data, without leaving the dashboard.`

Author: The reading flow is terrible if we keep it at the beginning of the sentence, so that's why I've opted to move it to the end. Usually I would follow UI order, but I don't think it'll harm comprehension here. 

And I'm reading the sentence again, and I think it's fine. Take another look maybe?

Reviewer: yeah lets keep it this way 👍

> In the **Dashboard controls** section of the sidebar in add mode, you can add filters and group by keys, variables, annotation queries, and dashboard links without leaving the dashboard.

Reviewer: I think it should be _edit_ mode

also, what if we said something like 

```
In the **Dashboard controls** section of the sidebar in edit mode, you can add controls to filter and group data, as well as add variables, annotation queries, and dashboard links without leaving the dashboard.
```

trying to avoid `filters and group by keys`

Author: I actually mean add mode. The **Dashboard controls** section only appears in the sidebar in this way when you've clicked the **Add new element** icon and the title of the sidebar is "Add". But calling it out as a UI name is tricky because it's not always the name of the sidebar. Ultimately, I think I'm overcomplicating it, and I'm just gonna take that part out.

```suggestion
In the **Dashboard controls** section of the sidebar, you can add variables, annotation queries, dashboard links, and controls to filter and group data, without leaving the dashboard.
```

## Filter and group by

<!-- vale Grafana.Spelling = NO -->

{{< admonition type="note" >}}
Filter and group by is currently in public preview.
Grafana Labs offers limited support, and breaking changes might occur prior to the feature being made generally available.

> This feature replaces ad hoc filters and extends them by adding grouping for Prometheus and Loki data sources.

Reviewer: it does not really replace adhoc, it extends it by adding grouping
plus we renamed adhoc -> filters

so not sure if it's a good call to say  we "replaced", but up to you

Author: ```suggestion
This feature renames _ad hoc filters_ to simply _filters_ and extends them by adding grouping for Prometheus and Loki data sources.
```

However, in the dashboard schema, it is still referred to as `"kind": "AdhocVariable"`.
To use this feature, enable the `dashboardUnifiedDrilldownControls` feature toggle in your Grafana configuration file.

For information on ad hoc filters, refer to the [Variables documentation](http://grafana.com/docs/grafana/<GRAFANA_VERSION>/visualizations/dashboards/variables/add-template-variables/#add-ad-hoc-filters).
While that documentation reflects the generally available experience, the information applies to the filter and group by feature.
{{< /admonition >}}

<!-- vale Grafana.Spelling = YES -->

The filter and group by is one of the most complex and flexible variable options available.

> Instead of creating a variable for each dimension by which you want to filter, it automatically creates variables (key/value pairs) for all the dimensions returned by your data source query.

Reviewer: `it automatically creates variables`

this is not exactly correct, thre's no automatically created variables, users need to create on filter (variable), those pull available key/values which the user can select and then we would automatically update queries based on that

 more like 

```Instead of creating a variable for each dimension by which you want to filter, they automatically query your data source for available dimensions and let users add or remove filters and groupings on the fly.```

> Instead of creating a variable for each dimension by which you want to filter, it automatically creates variables (key/value pairs) for all the dimensions returned by your data source query.

Author: I want to avoid the phrase "on the fly" because it's an idiom, which we avoid in the docs. Hopefully this is a good replacement.

```suggestion
Instead of creating a variable for each dimension by which you want to filter, they automatically query your data source for available dimensions and let users add or remove filters and groupings on the dashboard dynamically.
```

This allows you to quickly apply filters dashboard-wide.

> The group by function allows you to then group data by keys, letting you further narrow your scope.

Reviewer: group by does not narrow your scope, more like break/split your data into separate groups

Author: ```suggestion
The group by function allows you to then group data by keys, letting you split it up.
```

> Group by functionality works when you have queries that are aggregators, such as `sum(your_metric_here)`.

Reviewer: saying `only works with`  sounds a bit too strong
it's more like group by is meaningful with aggregation. You can still use thought

```
Group by is typically used with aggregation queries, such as sum(your_metric_here), to split aggregated results by selected dimensions.
```

Author: ```suggestion
Group by is typically used with aggregation queries, such as `sum(your_metric_here)`, to split aggregated results by the selected dimensions.
```

> Then, you can filter further with panel-level filters that let you drill down into your data.

Reviewer: not sure why we say `panel-level` here?

Author: ```suggestion
Then, you can use filters within panels to filter data in or out, drilling down further into the data.
```

> The filter and group by feature lets you add label/value filters that are automatically added to all metric queries that use the specified data source.

Reviewer: not necessarily metrics, you can use it with loki for example
I'd drop `metric`

Author: ```suggestion
The filter and group by feature lets you add label/value filters that are automatically added to all queries that use the specified data source.
```

Unlike other variables, you don't use these filters in queries.

> Instead, you use them to write filters for existing queries.

Reviewer: I'd add something about they get applied to the queries automatically

Author: Isn't that kind of what line 48 is saying? I'm wondering if it's redundant to add it on line 50.

```suggestion
Instead, you use them to write filters for existing queries and they're automatically applied to the queries.
```

Reviewer: yeah you're right, lets keep it as is

   | Use static key dimensions | To provide the filter dimensions as comma-separated values (CSV), toggle the switch on, and then enter the values in the space provided. Optional.                                                                                                                                                                                   |
   | Allow custom values       | Toggle the switch on to allow dashboard users to add custom values to the filter and group by lists. Optional.                                                                                                                                                                                                                       |

1. Click **Save**.
1. Enter an optional description of your dashboard changes, and then click **Save**.
1. Click **Exit edit**.

Now you can filter and group data on the dashboard.

> You can remove and reset default filters and group by selections, and see your recent ones:

Reviewer: `default filters and group by selections` -> `default filters and groupings`?

Author: ```suggestion
You can remove and reset default filters and groupings, and see your recent ones:
```

{{< figure src="/media/docs/grafana/dashboards/screenshot-reset-default-v13.0.png" max-width="500px" alt="Dashboard with the filters and group by selections" caption="Reset default filters and group by selections" >}}

{{< figure src="/media/docs/grafana/screenshot-filters-group-recent-v13.0.png" max-width="500px" alt="Dashboard with the filters and group by selections" caption="Recent filter and group by settings" >}}

> To see every active filter and group by across the dashboard all at once, click the filter icon in the toolbar to open an overview.

Reviewer: `grouping` as well?

Author: ```suggestion
To see every active filter and grouping across the dashboard all at once, click the filter icon in the toolbar to open an overview.
```

{{< figure src="/media/docs/grafana/dashboards/screenshot-panel-groupby-v13.0.png" max-width="550px" alt="Group by control on a panel" >}}

This can be helpful when you're working with a panel that's far away from the dashboard controls.
Your selection is applied to the all the panels in the dashboard with the same data source.

You can also further filter a time series panel, which allows you to drill down further into your data.
After setting your group by and splitting your data, click on a series in a panel and click `Filter on this value` or `Filter out this value` which will filter by the labels found on that series, which are related to the set group by values.

To enable this functionality, you need to add one or more overrides for the panel.
In the following example, the override:

- Adds a regular expression, so that all fields are filterable
- Enables the **Ad-hoc filterable** switch

{{< figure src="/media/docs/grafana/dashboards/screenshot-panel-filter-override-v13.0.png" max-width="400px" alt="Field override making all fields filterable" >}}

However, you can create overrides to address specific fields.

> Be sure to also set your data as filterable by either returning the dataframe with the appropriate `filterable` property on the desired fields

Reviewer: not sure about this sentence.. do you mean the datasource can set `filterable` on the dataframe?  I feel like we're mixing two things: users that can set `filterable` using overrides and datasources that can have it set but default

Author: Whoops! I took it straight out of some draft docs that Victor did and forgot to edit that, lol.
```suggestion
You can also do this programmatically by returning the data frame with the appropriate `filterable` property on the desired fields.
```

With the override in place, you can click a series on a time series panel and filter it in or out.
The new filter is shown in the dashboard filter control and the it's applied to the whole dashboard.

> <!-- TODO: Add screenshot here when feature is working properly -->

Author: @kristinademeshchik - I don't know what I'm doing wrong, but I can't get this to work to get a screenshot. Let's touch base and hopefully you can help me figure it out.

Reviewer: you mean this view?

<img width="410" height="265" alt="Image" src="https://github.com/user-attachments/assets/c1e20fc0-11c2-4250-b78c-285fd844da5b" />

Author: Yes! Can you take a screenshot of that but with the whole panel showing?

Reviewer: like this?

<img width="848" height="401" alt="Image" src="https://github.com/user-attachments/assets/99b24e02-e858-41af-9d8d-5ffc5bfaeeff" />

I can also help you make it work in your dashboard if you want

Author: This is excellent! I'll check in with you on Monday as to why I couldn't get it working.
